### PR TITLE
sync with english version for Inheritance and the prototype chain

### PR DESCRIPTION
**Pull request by yin1999**
### Description

sync with english version for Inheritance and the prototype chain

### Motivation

The description about `[[Prototype]]` and `__proto__` is a bit confusing.

### CODE REVIEW

File: files/zh-cn/web/javascript/inheritance_and_the_prototype_chain/index.md [hunk 1]
```diff
@@ -5,221 +5,439 @@ slug: Web/JavaScript/Inheritance_and_the_prototype_chain
 
 {{jsSidebar("Advanced")}}
 
-对于使用过基于类的语言 (如 Java 或 C++) 的开发者们来说，JavaScript 实在是有些令人困惑 —— JavaScript 是动态的，本身不提供一个 `class` 的实现。即便是在 ES2015/ES6 中引入了 `class` 关键字，但那也只是语法糖，JavaScript 仍然是基于原型的。
+对于使用过基于类的语言（如 Java 或 C++）的开发者来说，JavaScript 实在是有些令人困惑——JavaScript 是动态的且没有静态类型。
 
-当谈到继承时，JavaScript 只有一种结构：对象。每个实例对象（object）都有一个私有属性（称之为 \_\_proto\_\_）指向它的构造函数的原型对象（**prototype**）。该原型对象也有一个自己的原型对象（\_\_proto\_\_），层层向上直到一个对象的原型对象为 `null`。根据定义，`null` 没有原型，并作为这个**原型链**中的最后一个环节。
+当谈到继承时，JavaScript 只有一种结构：对象。每个对象（object）都有一个私有属性指向另一个名为原型（**prototype**）的对象。原型对象也有一个自己的原型，层层向上直到一个对象的原型为 `null`。根据定义，`null` 没有原型，并作为这个**原型链**（prototype chain）中的最后一个环节。可以改变原型链中的任何成员，甚至可以在运行时换出原型，因此 JavaScript 中不存在[静态分派](https://en.wikipedia.org/wiki/Static_dispatch)的概念。
```
Comment (yin1999) [inline files/zh-cn/web/javascript/inheritance_and_the_prototype_chain/index.md:10]:
```suggestion
当谈到继承时，JavaScript 只有一种结构：对象。每个对象（object）都有一个私有属性指向另一个名为**原型**（prototype）的对象。原型对象也有一个自己的原型，层层向上直到一个对象的原型为 `null`。根据定义，`null` 没有原型，并作为这个**原型链**（prototype chain）中的最后一个环节。可以改变原型链中的任何成员，甚至可以在运行时换出原型，因此 JavaScript 中不存在[静态分派](https://en.wikipedia.org/wiki/Static_dispatch)的概念。
```

File: files/zh-cn/web/javascript/inheritance_and_the_prototype_chain/index.md [hunk 2]
```diff
@@ -5,221 +5,439 @@ slug: Web/JavaScript/Inheritance_and_the_prototype_chain
...
+};
+
+// { a: 1, b: 2 } ---> { b: 3, c: 4 } ---> { d: 5 } ---> Object.prototype ---> null
+
+console.log(o.d); // 5
+```
+
+### 继承“方法”
+
+JavaScript 并没有其他基于类的语言所定义的“[方法](/en-US/docs/Glossary/Method)”。在 JavaScript 中，任何函数都被可以添加到对象上作为对象的属性。函数的继承与其他的属性继承没有差别，包括上面的“属性遮蔽”（这种情况相当于其他语言的*方法重写*）。
```
Comment (yin1999) [inline files/zh-cn/web/javascript/inheritance_and_the_prototype_chain/index.md:97]:
```suggestion
JavaScript 并没有其他基于类的语言所定义的“[方法](/en-US/docs/Glossary/Method)”。在 JavaScript 中，任何函数都被可以添加到对象上作为其属性。函数的继承与其他属性的继承没有差别，包括上面的“属性遮蔽”（这种情况相当于其他语言的*方法重写*）。
```

File: files/zh-cn/web/javascript/inheritance_and_the_prototype_chain/index.md [hunk 3]
```diff
@@ -5,221 +5,439 @@ slug: Web/JavaScript/Inheritance_and_the_prototype_chain
...
+console.log(child.method()); // 5
+// 因为 child 现在拥有“value”属性，“this.value”现在表示
+// child.value
 ```
 
-代码来源链接：<https://repl.it/@khaled_hossain_code/prototype>
+## 构造函数
 
-给对象设置属性会创建自有属性。获取和设置行为规则的唯一例外是当继承的属性带有 [getter 或 setter](/zh-CN/docs/Web/JavaScript/Guide/Working_with_objects#定义_getter_与_setter) 时。
+原型的强大之处在于，如果一组属性应该出现在每一个实例上，我们可以重用它们——尤其是对于方法。假设我们要创建多个盒子，其中每一个盒子都是一个对象，包含一个可以通过 `getValue` 函数访问的值。一个简单的实现可能是：
```
Comment (yin1999) [inline files/zh-cn/web/javascript/inheritance_and_the_prototype_chain/index.md:134]:
```suggestion
原型的强大之处在于，如果一组属性应该出现在每一个实例上，那我们就可以重用它们——尤其是对于方法。假设我们要创建多个盒子，其中每一个盒子都是一个对象，包含一个可以通过 `getValue` 函数访问的值。一个简单的实现可能是：
```

File: files/zh-cn/web/javascript/inheritance_and_the_prototype_chain/index.md [hunk 4]
```diff
@@ -5,221 +5,439 @@ slug: Web/JavaScript/Inheritance_and_the_prototype_chain
...
+```js-nolint
+const boxes = [
+  { value: 1, getValue() { return this.value; } },
+  { value: 2, getValue() { return this.value; } },
+  { value: 3, getValue() { return this.value; } },
+];
+```
 
-JavaScript 并没有其他基于类的语言所定义的“方法”。在 JavaScript 里，任何函数都可以添加到对象上作为对象的属性。函数的继承与其他的属性继承没有差别，包括上面的“属性遮蔽”（这种情况相当于其他语言的方法重写）。
+这是次优的，因为每一个实例都有自己的，做相同事情的函数属性，这是冗余且不必要的。相反，我们可以将 `getValue` 移动到所有盒子的 `[[Prototype]]` 上：
```
Comment (yin1999) [inline files/zh-cn/web/javascript/inheritance_and_the_prototype_chain/index.md:144]:
```suggestion
这是不够好的，因为每一个实例都有自己的，做相同事情的函数属性，这是冗余且不必要的。相反，我们可以将 `getValue` 移动到所有盒子的 `[[Prototype]]` 上：
```

File: files/zh-cn/web/javascript/inheritance_and_the_prototype_chain/index.md [hunk 5]
```diff
@@ -5,221 +5,439 @@ slug: Web/JavaScript/Inheritance_and_the_prototype_chain
...
+};
+
+const boxes = [
+  { value: 1, __proto__: boxPrototype },
+  { value: 2, __proto__: boxPrototype },
+  { value: 3, __proto__: boxPrototype },
+];
+```
+
+这样，所有盒子的 `getValue` 方法都会引用相同的函数，降低了内存使用率。但是，手动绑定每个对象创建的 `__proto__` 仍旧非常不方便。这时，我们就可以使用*构造函数*，它会自动为每个构造的对象设置 `[[Prototype]]`。构造函数是使用 [[`new`](/en-US/docs/Web/JavaScript/Reference/Operators/new) 调用的函数。
```
Comment (yin1999) [inline files/zh-cn/web/javascript/inheritance_and_the_prototype_chain/index.md:160]:
```suggestion
这样，所有盒子的 `getValue` 方法都会引用相同的函数，降低了内存使用率。但是，手动绑定每个对象创建的 `__proto__` 仍旧非常不方便。这时，我们就可以使用*构造函数*，它会自动为每个构造的对象设置 `[[Prototype]]`。构造函数是使用 [`new`](/en-US/docs/Web/JavaScript/Reference/Operators/new) 调用的函数。
```

File: files/zh-cn/web/javascript/inheritance_and_the_prototype_chain/index.md [hunk 6]
```diff
@@ -5,221 +5,439 @@ slug: Web/JavaScript/Inheritance_and_the_prototype_chain
...
+
+// 使用 Box() 构造函数创建的所有盒子都将具有的属性
+Box.prototype.getValue = function () {
+  return this.value;
+};
+
+const boxes = [new Box(1), new Box(2), new Box(3)];
+```
+
+我们说 `new Box(1)` 是通过 `Box` 构造函数创建的一个*实例*。`Box.prototype` 与我们之前创建的 `boxPrototype` 并无太大区别——它只是一个普通的对象。通过构造函数创建的每一个实例都会自动将构造函数的 [`prototype`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/prototype) 属性作为其 `[[Prototype]]`。即，`Object.getPrototypeOf(new Box()) === Box.prototype`。`Constructor.prototype` 默认具有一个自由属性：[`constructor`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/constructor)，它引用了构造函数本身。即，`Box.prototype.constructor === Box`。这允许我们从任何实例中访问原始构造函数。
```
Comment (yin1999) [inline files/zh-cn/web/javascript/inheritance_and_the_prototype_chain/index.md:176]:
```suggestion
我们说 `new Box(1)` 是通过 `Box` 构造函数创建的一个*实例*。`Box.prototype` 与我们之前创建的 `boxPrototype` 并无太大区别——它只是一个普通的对象。通过构造函数创建的每一个实例都会自动将构造函数的 [`prototype`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/prototype) 属性作为其 `[[Prototype]]`。即，`Object.getPrototypeOf(new Box()) === Box.prototype`。`Constructor.prototype` 默认具有一个自由属性：[`constructor`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/constructor)，它引用了构造函数本身。即，`Box.prototype.constructor === Box`。这允许我们在任何实例中访问原始构造函数。
```

File: files/zh-cn/web/javascript/inheritance_and_the_prototype_chain/index.md [hunk 7]
```diff
@@ -5,221 +5,439 @@ slug: Web/JavaScript/Inheritance_and_the_prototype_chain
...
+// 请不要这样做——使用 Object.setPrototypeOf 来修改它
+Derived.prototype = Object.create(Base.prototype);
 ```
 
-在控制台显示的 JavaScript 代码块中，我们可以看到 doSomething 函数的一个默认属性 prototype。而这段代码运行之后，控制台应该显示类似如下的结果：
+## 检查原型：更深入的研究
+
+让我们来仔细看看幕后发生了什么。
+
+In JavaScript, as mentioned above, functions are able to have properties. All functions have a special property named `prototype`. Please note that the code below is free-standing (it is safe to assume there is no other JavaScript on the webpage other than the below code). For the best learning experience, it is highly recommended that you open a console, navigate to the "console" tab, copy-and-paste in the below JavaScript code, and run it by pressing the Enter/Return key. (The console is included in most web browser's Developer Tools. More information is available for [Firefox Developer Tools](https://firefox-source-docs.mozilla.org/devtools-user/index.html), [Chrome DevTools](https://developer.chrome.com/docs/devtools/), and [Edge DevTools](https://docs.microsoft.com/archive/microsoft-edge/legacy/developer/).)
```
Comment (yin1999) [inline files/zh-cn/web/javascript/inheritance_and_the_prototype_chain/index.md:322]:
```suggestion
```

File: files/zh-cn/web/javascript/inheritance_and_the_prototype_chain/index.md [hunk 8]
```diff
@@ -5,221 +5,437 @@ slug: Web/JavaScript/Inheritance_and_the_prototype_chain
...
+
+// 使用 Box() 构造函数创建的所有盒子都将具有的属性
+Box.prototype.getValue = function () {
+  return this.value;
+};
+
+const boxes = [new Box(1), new Box(2), new Box(3)];
+```
+
+我们说 `new Box(1)` 是通过 `Box` 构造函数创建的一个*实例*。`Box.prototype` 与我们之前创建的 `boxPrototype` 并无太大区别——它只是一个普通的对象。通过构造函数创建的每一个实例都会自动将构造函数的 [`prototype`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Function/prototype) 属性作为其 `[[Prototype]]`。即，`Object.getPrototypeOf(new Box()) === Box.prototype`。`Constructor.prototype` 默认具有一个自由属性：[`constructor`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Object/constructor)，它引用了构造函数本身。即，`Box.prototype.constructor === Box`。这允许我们在任何实例中访问原始构造函数。
```
Comment (jasonren0403) [inline files/zh-cn/web/javascript/inheritance_and_the_prototype_chain/index.md:176]:
```suggestion
我们说 `new Box(1)` 是通过 `Box` 构造函数创建的一个*实例*。`Box.prototype` 与我们之前创建的 `boxPrototype` 并无太大区别——它只是一个普通的对象。通过构造函数创建的每一个实例都会自动将构造函数的 [`prototype`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Function/prototype) 属性作为其 `[[Prototype]]`。即，`Object.getPrototypeOf(new Box()) === Box.prototype`。`Constructor.prototype` 默认具有一个自有属性：[`constructor`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Object/constructor)，它引用了构造函数本身。即，`Box.prototype.constructor === Box`。这允许我们在任何实例中访问原始构造函数。
```

File: files/zh-cn/web/javascript/inheritance_and_the_prototype_chain/index.md [hunk 9]
```diff
@@ -5,221 +5,437 @@ slug: Web/JavaScript/Inheritance_and_the_prototype_chain
...
+// 以作为其 `[[Prototype]]` 的新对象
+// 请不要这样做——使用 Object.setPrototypeOf 来修改它
+Derived.prototype = Object.create(Base.prototype);
+```
+
+## 检查原型：更深入的研究
+
+让我们来仔细看看幕后发生了什么。
+
+如上所属，在 JavaScript 中，函数可以拥有属性。所有函数都有一个名为 `prototype` 的特殊属性。请注意，下面的代码是独立的（出于严谨，假设页面没有其他的 JavaScript 代码）。为获得最佳的学习体验，强烈建议你打开控制台，进入“console”标签页，复制并粘贴以下 JavaScript 代码，然后按回车键运行。（大多数 web 浏览器的开发者工具中都包含控制台。请参阅 [Firefox 开发者工具](https://firefox-source-docs.mozilla.org/devtools-user/index.html)、[Chrome 开发者工具](https://developer.chrome.com/docs/devtools/)和 [Edge 开发者工具](https://docs.microsoft.com/archive/microsoft-edge/legacy/developer/)，以了解详情。）
```
Comment (jasonren0403) [inline files/zh-cn/web/javascript/inheritance_and_the_prototype_chain/index.md:322]:
```suggestion
如上所述，在 JavaScript 中，函数可以拥有属性。所有函数都有一个名为 `prototype` 的特殊属性。请注意，下面的代码是独立的（出于严谨，假设页面没有其他的 JavaScript 代码）。为获得最佳的学习体验，强烈建议你打开控制台，进入“console”标签页，复制并粘贴以下 JavaScript 代码，然后按回车键运行。（大多数 web 浏览器的开发者工具中都包含控制台。请参阅 [Firefox 开发者工具](https://firefox-source-docs.mozilla.org/devtools-user/index.html)、[Chrome 开发者工具](https://developer.chrome.com/docs/devtools/)和 [Edge 开发者工具](https://docs.microsoft.com/archive/microsoft-edge/legacy/developer/)，以了解详情。）
```

File: files/zh-cn/web/javascript/inheritance_and_the_prototype_chain/index.md [hunk 10]
```diff
@@ -228,86 +444,150 @@ doSomething.prototype.prop: undefined
...
+      >对象初始化器</a>中使用 <code>__proto__</code> 键的优缺点
+  </caption>
+  <tbody>
+    <tr>
+      <th scope="row">优点</th>
+      <td>
+        被所有的现代引擎所支持。将 <code>__proto__</code>
+        属性指向非对象的值只会被忽略，而非抛出异常。与
+        {{jsxref("Object/proto", "Object.prototype.__proto__")}} setter 相反，对象字面量初始化器中的 <code>__proto__</code>
+        是标准化，被优化的。甚至可以比 {{jsxref("Object.create")}}更高效。在创建对象时声明额外的自有属性比 {{jsxref("Object.create")}}
```
Comment (jasonren0403) [inline files/zh-cn/web/javascript/inheritance_and_the_prototype_chain/index.md:489]:
```suggestion
        是标准化，被优化的。甚至可以比 {{jsxref("Object.create")}} 更高效。在创建对象时声明额外的自有属性比 {{jsxref("Object.create")}}
```

**Review by jasonren0403 — COMMENTED** [review]
一些小修改

**Review by jasonren0403 — APPROVED** [review]

**Review by jasonren0403 — APPROVED** [review]

**Pull request merged**